Add Spark Scala Dataproc Wordcount Example

waprin opened the pull request:
This is an update of the Spark example previously in this repo. Now includes instructions for packaging and running on local Spark as well as on Cloud Dataproc.

Currently has a test that runs with local Spark and deployed Bigtable. TODO is add full e2e test.

sbt seems more popular in spark world but was having problems with it so using Maven for now, TODO to switch.

waprin commented:
@nkashy1 please review when you get chance next week can't seem to assign you

waprin commented:
@sduskis @nkashy1 ping.

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+The Spark job is assembled into a fat jar with all its dependencies. To build, run
+
+   mvn assembly:assembly
```
Add one more space indent to make this code-formatted.

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+Now submit your job to Cloud Dataproc:
+
+    gcloud dataproc jobs submit spark --cluster spark-cluster \
+    --class com.example.bigtable.spark.wordcount.WordCount  \
```
Indent each continuation line by 4 spaces for readability?

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+cluster.
+
+    spark-submit --master local --class com.example.bigtable.spark.wordcount.WordCount \
+    target/cloud-bigtable-dataproc-spark-wordcount-0.1-jar-with-dependencies.jar \
```
Indent each continuation line by 4 spaces for readability?

waprin replied:
done

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+### Create a Google Cloud Storage bucket
+
+For deployed jobs on Cloud Dataproc, it's easiest to read a file from Google
+Cloud Storage. You can use `gsutil` to create the bucket.
```
`s/bucket./bucket:/`

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+## Prerequisites
+
+1. A [Google Cloud project](console.cloud.google.com) with billing enabled. Please
```
https://console.cloud.google.com

Otherwise, it'll be relative to current URL, and hence, a 404.

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+## Overview
+
+This sample demonstrates how to use  [Google Cloud Dataproc](http://cloud.google.com/dataproc), which
```
`https://`

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+This sample demonstrates how to use  [Google Cloud Dataproc](http://cloud.google.com/dataproc), which
+provides a managed [Apache Spark](https://spark.apache.org/) environment with
+[Google Cloud Bigtable](http://cloud.google.com/bigtable/docs).
```
`https://`

waprin replied:
done

mbrukman commented on spark/wordcount/src/test/scala/com/example/bigtable/spark/wordcount/SparkSpec.scala:
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
```
If this is written by Google, it should have the same license header as the other file, not explicitly licensed to the ASF, but just licensed under Apache 2.0.

`WordCountIT.scala` seems to have the correct header.

waprin replied:
done

mbrukman commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
```
If this is written by Google, it should have the same license header as the other file, not explicitly licensed to the ASF, but just licensed under Apache 2.0.

`WordCountIT.scala` seems to have the correct header.

waprin replied:
done

pmkc commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+    gcloud dataproc jobs submit spark --cluster spark-cluster \
+    --class com.example.bigtable.spark.wordcount.WordCount  \
+    --jar target/cloud-bigtable-hello-world-1.0-SNAPSHOT-jar-with-dependencies.jar \
```
Technically you need to use --jars instead --jar implies it's providing the main class.

waprin replied:
done

pmkc commented on spark/wordcount/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client</artifactId>
+            <version>2.7.2</version>
```
May as well parameterize hadoop.version

sduskis replied:
Direct dependencies on hadoop-client, hadoop-common, hbase-client and even netty-tcnative-boringssl-static should not be necessary in the 1.0.0-pre2 version.

waprin replied:
done

waprin replied:
done

pmkc commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,141 @@
+
+    // Create a per-partition connection to ensure each node has a
+    // connection (partitions are on at most 1 node).
+    wordCounts.foreachPartition {
```
I would encourage you to use saveAsNewHadoopDataset as I do here:
https://github.com/pmkc/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/src/main/scala/WordCount.scala#L72

waprin replied:
Unfortunately your PR doesn't work due to ALPN issues.

waprin replied:
Ok, once got some through some dependency issues got your approach working. Only big disadvantage of this approach is currently if you screw up small configuration you get a really cryptic error

pmkc commented on spark/wordcount/pom.xml:
```diff
@@ -0,0 +1,154 @@
+
+        <!-- hadoop-client and hadoop-common are not used directly. They are
+        included here to workaround an issue with mismatched versions of Guava
+        being needed by hbase-client and spark -->
```
Shading is often necessary on the Spark Classpath, but I'm not sure in this case.
You could probably just exclude Guava from the Spark dependency, because Spark shades it's internal Guava.

waprin replied:
lots of this stuff got fixed with bigtable-hbase version change

pmkc commented on spark/wordcount/pom.xml:
```diff
@@ -0,0 +1,154 @@
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_2.11</artifactId>
+            <version>${spark.version}</version>
+            <!---<scope>provided</scope>-->
```
I would think Spark should be provided

waprin replied:
done

pmkc commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+1. You have the [Google Cloud SDK](https://cloud.google.com/sdk/) installed.
+
+1. You have [Scala](https://www.scala-lang.org/) installed.
```
Local Scala installations are never necessary with Spark (though they can help for debugging).

waprin replied:
added note

sduskis commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+    gcloud dataproc clusters delete spark-cluster
+
+
+## Running tests
```
We might be able to use a Cloud Bigtable emulator here, but that's not a requirement.

waprin replied:
Oh I should use that, but will do another PR. Is there a way to spin it up progrmatically?

sduskis commented on spark/wordcount/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    <version>0.1</version>
+
+    <properties>
+        <bigtable.version>0.9.7.1</bigtable.version>
```
bigtable.version 1.0.0-pre2 is preferred.

waprin replied:
done

sduskis commented on spark/wordcount/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        </dependency>
+        <dependency>
+            <groupId>com.google.cloud.bigtable</groupId>
+            <artifactId>bigtable-hbase-1.1</artifactId>
```
The bigtable-hbase-1.x-hadoop artifact should be used here, if you upgrade to the 1.0.0-pre2 version

waprin replied:
done

sduskis commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,141 @@
+  val ColumnNameBytes = Bytes.toBytes("Count")
+
+  def createConnection(ProjectId: String, InstanceID: String): Connection = {
+    BigtableConfiguration.connect(ProjectId, InstanceID)
```
There are some settings that work well for Batch systems.  [Here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/bigtable-client-1.0.0-pre2/bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableConfiguration.java#L162)'s how we do it for Dataflow.

waprin replied:
Thanks, now not relevant since using saveAsNewAPIHadoop dataset

sduskis commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,141 @@
+        val partitionConnection = createConnection(projectId, instanceId)
+        val table = TableName.valueOf(tableName)
+        val mutator = partitionConnection.getBufferedMutator(table)
+        try {
```
Java and Scala have a notion of try-with-resources, which makes this type of code (mutator creation and closing) a bit easier.  [Here](https://www.phdata.io/try-with-resources-in-scala/)'s one article about it.

waprin replied:
Thanks for tip, not relevant here anymore

sduskis commented on spark/wordcount/src/test/scala/com/example/bigtable/spark/wordcount/WordCountIT.scala:
```diff
@@ -0,0 +1,115 @@
+
+    val f = fixture
+
+    val mutator = f.connection.getBufferedMutator(f.tableName)
```
try-with-resources would make this a bit nicer.

waprin replied:
most of this code got removed, and in the case of tests, cleaning up in the afterEach/afterAll seems cleaner to me.

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+If you don't already have a Cloud Bigtable instance, create one
+
+     gcloud bigtable instances create test-instance --cluster test-cluster --cluster-zone us-east1-b --cluster-num-nodes 3
```
Unfortunately, this still needs to be `gcloud beta bigtable`; here and elsewhere.

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+## Build the jar
+
+The Spark job is assembled into a fat jar with all its dependencies. To build, run
```
s/all its/all of its/

mbrukman replied:
s/run$/run:/

waprin replied:
done

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+    your-project-id your-bigtable-instance-id wordcount-scratch \
+    src/test/resources/countme.txt
+
+The job will create the table specified (here, `wordcount`) if it doesn't already exist.
```
The table above appears to be `wordcount-scratch`, not `wordcount`; is that correct?

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+
+    spark-submit --master local --class com.example.bigtable.spark.wordcount.WordCount \
+    target/cloud-bigtable-dataproc-spark-wordcount-0.1-jar-with-dependencies.jar \
+    your-project-id your-bigtable-instance-id wordcount-scratch \
```
Would it make sense to use the same strategy for template parameters as the docs, i.e., using `[PROJECT_ID]`, `[BIGTABLE_INSTANCE]`, and `[BIGTABLE_TABLE]` to clarify they are variables, not verbatim parameters?

waprin replied:
done

mbrukman commented on spark/wordcount/README.md:
```diff
@@ -0,0 +1,102 @@
+delete your resources.
+
+    gsutil rm your-bucket-id
+    gcloud bigtable instances delete test-instance
```
`gcloud beta bigtable` here as well, for the time being.

waprin replied:
done

waprin commented:
@pmkc thanks for review, w.r.t. some suggested Maven and hadoop.writeAsDataAPI, your PR did not work, my suggestion is we merge as is, make sure we have CI setup correctly, if you have a better way of doing things put a PR on top that runs correctly with spark-submit and dataproc submit. Now we're a bit deadlocked because we don't have working samples, I have a working sample now, and getting blocked by suggestions that don't work and it's not a good use of time to spin wheels on it too long.

waprin commented:
@sduskis @pmkc @mbrukman big thanks again for review and all help. Many of the dependency issues do seem to have been improved once using correct version of bigtable-hbase. I switched to saveAsNewAPIHadoop though as noted misconfigurations give you a pretty confusing/generic error message.

Think I addressed all comments so take another look whenever. Also I should prob just be consistent and change this directory to `scala`, but I think its easier to do in a separate PR since file renames tend to confuse the code review.

waprin commented:
unfortunately got ahead of myself and still getting error when deploying to dataproc:

    java.lang.IllegalArgumentException: Can not create a Path from a null string

investigating...

waprin commented:
Ah found Spark JIRA issue with workaround so works now.

waprin commented:
@sduskis @mbrukman ping

sduskis commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,121 @@
+
+    var conf = BigtableConfiguration.configure(
+      projectId, instanceId)
+    conf.set(TableInputFormat.INPUT_TABLE, tableName)
```
Can you please add the Dataflow configuration options here?

There are some settings that work well for Batch systems.  [Here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/bigtable-client-1.0.0-pre2/bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableConfiguration.java#L162)'s how we do it for Dataflow.

waprin replied:
done, sorry I misunderstood when you asked the first time and thought that it was necessary when just using bufferedmutator but would already be done when using saveAsNewAPIHadoopDataset approach

sduskis commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,121 @@
+    conf.setInt(HConstants.HBASE_CLIENT_OPERATION_TIMEOUT, 60000)
+    conf.set(
+      "hbase.client.connection.impl",
+      "com.google.cloud.bigtable.hbase1_x.BigtableConnection")
```
Please remove this:

>    conf.set(
>      "hbase.client.connection.impl",
>      "com.google.cloud.bigtable.hbase1_x.BigtableConnection")

BigtableConfiguration.configure() already does the right thing, and will be more resilient overtime as we change the Connection class (like in hbase2).

waprin replied:
done

sduskis commented on spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala:
```diff
@@ -0,0 +1,154 @@
+
+    var conf = BigtableConfiguration.configure(
+      projectId, instanceId)
+    conf.set(TableInputFormat.INPUT_TABLE, tableName)
```
Can you please remove out this line:
` conf.set(TableInputFormat.INPUT_TABLE, tableName)`
`TableOutputFormat.OUTPUT_TABLE` is used, and not the `INPUT_TABLE` setting, since you're using a `TableOutputFormat` (in line 127).

waprin replied:
done

sduskis commented on spark/wordcount/pom.xml:
```diff
@@ -0,0 +1,141 @@
+            <artifactId>spark-core_2.11</artifactId>
+            <version>${spark.version}</version>
+        </dependency>
+        <dependency>
```
Is spark sql needed here?

waprin replied:
removed it

waprin commented:
@mbrukman just waiting on you to merge

waprin commented:
@mbrukman ping

waprin commented:
since I got solomon's approval and guessing misha's just busy, going to merge for now, let me know if there's any issues.
